### ISSUE TITLE
Investigate creating a low-level crate so that users can easily customize `zip`-crate.

### ISSUE THREAD

**Issue by xMAC94x** [zip-rs/zip2#204]
**Is your feature request related to a problem? Please describe.**
So this story for this issue originates from the veloren project, where we have a 300MB zip file, network hosted for the game and want to find a way to update client files without re-downloading the whole zip and unzipping it. We found out that due to the zip structure, we could use HTTP RANGE requests and just analyze the CentralDirectoryHeaders and check CRC mismatches locally. Somehow inject this in the zip-extracting process.

**Describe the solution you'd like**
The first thing I found out is that this is currently not possible with `zip`, and maybe also not the intended use-case of `zip`-crate.
However, it would be nice to have a `low-level` crate that shares `structs` with zip so that users of the `zip` crate maybe can interact with it, by constructing/reading the `low-level` stuff manually and passing it to `zip`.

Additionally, I looked at other crates `rc-zip`  and `zip-structs`, all have the issue that they have to redefine the [APPNOTE.TXT](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT) someway or the other. Making each crate incompatible with others.

Where network crates have `http` or even `url` in the world of `zip` there is no crate that reexports the simplest stuff for people to work with.

IMO such a crate is only worth it if it finds adoption. To find adoption it must be somewhat helpful for `zip` developers, on the other hand not restrictive to future development. It should include as less dependencies as necessary. Should not force anyone to use sync or async.

I looked at the APPNOTE.TXT and tried to implement a minimum example low-level crate. Spend a few days for: https://crates.io/crates/zip-core/0.0.2
Nothing extraordinary, defining 1:1 mappings of the spec (Central Directory, Local File Header, a few helpers for checking if signatures are correct, optional parsing (via a crate feature if needed), Structs that contain all valid CompressionMethods, all without dependencies by default and no-std.

I wanted to start a discussion, does it make sense to create such a low-level crate, that focuses on structs of the APPNOTE.TXT, optionally some parsing helpers and that's it. While the `zip` crate focuses on the hard stuff. figuring out which version works together with which field. Use the correct Encryption Algorithm, find the position of the EndOfCentralDirectoryRecord. Reading data from std::io::Read, etc...

Feel free to comment on this. Even if it's just a declination, the idea and crate is at an early stage and still can be shaped easily :)

**Describe alternatives you've considered**
The simplest alternative is to ignore this approach, say such a low-level crate is unhelpful, maybe not necessary at all and keep continuing like usual. However, making it harder for people to interact in the zip environment.

Best regards,
Marcel

**Comment by im7mortal** [comment]
A slightly different approach to customizing zip would be to use dependency injection.

I was investigating two issues: in one case, a user wanted to disable CRC checks for their reverse engineering tool; in another, a user wanted to customize AES writing to achieve deterministic results. In both cases, users suggested adding additional flags and parameters.

However, both scenarios could be resolved using dependency injection instead of introducing an ever-growing number of configuration flags. If a user needs specific behavior, they could reuse most of the zip crate’s functionality and replace only the relevant logic.

At the moment, I am investigating how to make the Read and Write operations more flexible. The Read side has four layers of Read traits, and the Write side similarly has four layers of Write traits. There are fixed implementations (for example, `std::io::BufWriter` ) as well as implementations that use static enum dispatch. Like:

```rust
impl<R: Read + ?Sized> Read for CryptoReader<'_, R> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        match self {
            CryptoReader::Plaintext(r) => r.read(buf),
            CryptoReader::ZipCrypto(r) => r.read(buf),
            #[cfg(feature = "aes-crypto")]
            CryptoReader::Aes { reader: r, .. } => r.read(buf),
        }
    }
}
```

In [this comment](
https://github.com/zip-rs/zip2/issues/170#issuecomment-3917300648), I explained how it would be possible to preserve static dispatch for the majority of use cases while still allowing custom logic to be injected when needed.

**Comment by im7mortal** [comment]
I’m starting to believe more and more that this is the correct way.

The idea is that the ZIP core library should be low-level, and it’s fine if it’s somewhat hard to use.

It should contain only the most essential functionality.

#### Read

- stream and seekable support
- reader object similar to io::Read, but usable in no_std. Maybe this will become available eventually:
- https://github.com/rust-lang/rust/issues/48331#issuecomment-3935875937  (This would also apply to Read and Seek.)
- No implementations of decryption, decompression, or CRC32 checksum. These should be provided via generics or dyn traits (yes dyn traits, I hope we can have devirtualization in the future https://github.com/rust-lang/rust/issues/68262 ). 
- internal structures (central directory, files) should be accessible in read mode, and possibly writable (maybe unsafe if needed).
- file iteration (for non-seekable streams) should support iterators with [guaranteed zero-cost abstraction](https://turbopuffer.com/blog/zero-cost).

#### Write

- stream and seekable support
- writer object
- No implementations of encryption, compression, or CRC32 checksum (same reasoning as in the read section)
- internal structures (central directory, files) accessible in read mode and possibly write mode (not fully sure about this yet)

#### Runtime

- runtime-agnostic design
- multiple runtimes supported in parallel
- no default runtime dependency

#### Codebase

- ideally formally verifiable
- easy to audit (especially crypto)
